src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -1715,7 +1715,36 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
       "The stateless mutator does not specify both 'argument_tags' and "
       "'return_tags'. Did you forget to add these tag lists to the mutator "
       "class? The class is listed as the first template parameter below.");
-  mutate_apply(F{}, box, std::forward<Args>(args)...);
+  return mutate_apply(F{}, box, std::forward<Args>(args)...);
+}
+
+namespace DataBox_detail {
+template <typename... MutateTags, typename BoxTags, typename... Args>
+SPECTRE_ALWAYS_INLINE constexpr void mutate_assign_impl(
+    // NOLINTNEXTLINE(readability-avoid-const-params-in-decls)
+    const gsl::not_null<DataBox<BoxTags>*> box,
+    tmpl::list<MutateTags...> /*meta*/, Args&&... args) noexcept {
+  static_assert(sizeof...(MutateTags) == sizeof...(args),
+                "The number of arguments passed to `mutate_assign` must be "
+                "equal to the number of tags passed.");
+  db::mutate<MutateTags...>(box, [&args...](const auto... box_args) noexcept {
+    // silence unused capture warnings
+    EXPAND_PACK_LEFT_TO_RIGHT((void)args);
+    EXPAND_PACK_LEFT_TO_RIGHT((*box_args = std::forward<Args>(args)));
+  });
+}
+}  // namespace DataBox_detail
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Perform a mutation to the \ref DataBoxGroup `box`, assigning the
```

Comment:
This is just a DataBox utility, not inherently related to initialization. I don't see a reason we should discourage using this function. For example, I can see it be useful to move data from an inbox into the DataBox in a receive action.

Reply:
Yea, we don't want anyone to ever do that... Basically, this utility will cause people to blindly never reuse memory allocations and then we have to waste time fixing lots of code. It's better to prevent a bug than to fix one.

Reply:
I'm not sure I understand - how else would you move data from the inbox into the DataBox?

Reply:
`db::mutate`. My concern is that people will start using `mutate_assign` everywhere, and in many places where you should be using `db::mutate`. The "it's less code I'll just use that" issue. This will mean we don't use re-use a lot of memory allocations and just adds more manual policing to the code review process. The number of times `mutate_assign` should be used outside of initialization is close to zero, so I'd rather eliminate an entire class of bugs than have to add yet another thing to the "Please remember to check this while reviewing code" list.

Reply:
ok